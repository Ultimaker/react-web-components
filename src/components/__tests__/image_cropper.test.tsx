// Copyright (c) 2018 Ultimaker B.V.
import * as React from 'react';
import { shallow } from 'enzyme';
import AvatarEditor from 'react-avatar-editor';
import RangeSlider from '../range_slider';
import ImageCropper from '../image_cropper';

describe('The Image Cropper component', () => {
    let props;
    const onImageChanged = jest.fn();
    const onCropCancel = jest.fn();
    const mockEditor = { getImageScaledToCanvas: () => ({ toDataURL: () => 'imageData' }) };

    beforeEach(() => {
        onImageChanged.mockReset();
        onCropCancel.mockReset();
        props = {
            onImageChanged,
            onCropCancel,
        };
    });

    const createWrapper = () => {
        const wrapper = shallow(<ImageCropper {...props} />);
        // @ts-ignore
        wrapper.instance()._editor = mockEditor;
        return wrapper;
    };

    it('should render', () => {
        const wrapper = createWrapper();
        expect(wrapper).toMatchSnapshot();
        expect(onImageChanged).not.toHaveBeenCalled();
        expect(onCropCancel).not.toHaveBeenCalled();
    });

    it('should give the avatar a border radius', async () => {
        props.shape = 'round';
        props.size = '18rem';
        props.borderSize = 20;
        const wrapper = createWrapper();
        expect(wrapper.find(AvatarEditor).props()).toEqual(expect.objectContaining({
            width: 140,
            height: 140,
            border: 20,
            borderRadius: 140,
        }));
    });

    it('should keep the range slider', async () => {
        const wrapper = createWrapper();
        expect(wrapper.find(RangeSlider).prop('value')).toEqual(1);
        expect(wrapper.find(AvatarEditor).prop('scale')).toEqual(1);
        wrapper.find(RangeSlider).prop('onChange')(0.5);
        expect(wrapper.find(RangeSlider).prop('value')).toEqual(0.5);
        expect(wrapper.find(AvatarEditor).prop('scale')).toEqual(0.5);
    });

    it('should keep the image position', async () => {
        const wrapper = createWrapper();
        const avatarProps: any = wrapper.find(AvatarEditor).props();
        expect(avatarProps.position).toEqual({ x: 0.5, y: 0.5 });
        avatarProps.onPositionChange({ x: 0, y: 1 });
        expect(wrapper.find(AvatarEditor).prop('position')).toEqual({ x: 0, y: 1 });
    });

    it('should parse the callback', async () => {
        const wrapper = createWrapper();
        const avatarProps: any = wrapper.find(AvatarEditor).props();
        avatarProps.onImageChange();
        // @ts-ignore
        wrapper.instance()._onImageChanged.flush();
        expect(onImageChanged).toHaveBeenCalledWith('imageData');
        expect(onCropCancel).not.toHaveBeenCalled();
    });
});
