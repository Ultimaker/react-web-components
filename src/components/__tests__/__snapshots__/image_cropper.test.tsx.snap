// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Image Cropper component should render 1`] = `
<div
  className="image-cropper--container"
>
  <d
    border={25}
    borderRadius={0}
    className="editor-canvas"
    color={
      Array [
        0,
        0,
        0,
        0.5,
      ]
    }
    disableBoundaryChecks={false}
    height={150}
    image="/images/Ultimaker-Resources.jpg"
    onImageChange={[Function]}
    onImageReady={[Function]}
    onLoadFailure={[Function]}
    onLoadSuccess={[Function]}
    onMouseMove={[Function]}
    onMouseUp={[Function]}
    onPositionChange={[Function]}
    position={
      Object {
        "x": 0.5,
        "y": 0.5,
      }
    }
    rotate={0}
    scale={1}
    style={Object {}}
    width={150}
  />
  <RangeSlider
    className="image-cropper--slider"
    max={2.5}
    min={1}
    onChange={[Function]}
    step={0.1}
    value={1}
  />
  <Button
    className="remove-image"
    linkToNewTab={false}
    onClickHandler={[MockFunction]}
    shape="pill"
    style="quiet"
    type="button"
  >
    <Icon(CrossIcon)
      color="white"
      size="sm"
    />
  </Button>
</div>
`;
